### PR TITLE
Implement refresh endpoint

### LINKED ISSUE #10: Implement refresh endpoint
An endpoint and relevant application logic should be implemented that will take a JWT access token along with a JWT refresh token, generate an updated JWT access token, and return it along with the JWT refresh token as an HTTP-only cookie.

### PR DESCRIPTION
**Pull request by VKTB**
## Description
Implements a `/refresh` `POST` endpoint and relevant application logic that takes a JWT access token along with a JWT refresh token, generates an updated JWT access token, and returns it along with the JWT refresh token as an HTTP-only cookie.

## Testing instructions
Add a set of instructions describing how the reviewer should test the code
- [ ] Review code
- [ ] Check Actions build
- [ ] Supply valid access token
- [ ] Supply invalid access token
- [ ] Supply invalid refresh token
- [ ] Supply no refresh token

## Agile board tracking
closes #10

### CODE REVIEW

File: ldap_jwt_auth/auth/jwt_hanlder.py [hunk 1]
```diff
@@ -44,6 +44,27 @@ def get_refresh_token(self) -> str:
...
+        :raises JWTRefreshError: If the JWT access token cannot be refreshed.
+        :return: JWT access token with an updated expiry time.
+        """
+        logger.info("Refreshing access token")
+        self.verify_token(refresh_token)
+        try:
+            payload = self._get_jwt_payload(access_token, {"verify_exp": False})
+            payload["exp"] = datetime.now(timezone.utc) + timedelta(
+                minutes=config.authentication.access_token_validity_minutes
+            )
```
Comment (joelvdavies) [inline ldap_jwt_auth/auth/jwt_hanlder.py:61]: Just some info on token rotation https://auth0.com/docs/secure/tokens/refresh-tokens/refresh-token-rotation I mentioned. No need here though.

File: ldap_jwt_auth/routers/refresh.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+
+@router.post(
+    path="/",
+    summary="Generate an updated JWT access token using the JWT refresh token",
+    response_description="A JWT access token",
+)
+def refresh_access_token(
+    jwt_handler: Annotated[JWTHandler, Depends(JWTHandler)],
+    access_token: Annotated[str, Query(description="The JWT access token to refresh")],
```
Comment (joelvdavies) [inline ldap_jwt_auth/routers/refresh.py:26]: This needs to be named `token` in the body for SciGateway (See https://github.com/ral-facilities/operationsgateway-api/blob/fb655d976e794e23a96828f6d2ad8cdef46d0141/operationsgateway_api/src/routes/auth.py#L18-L24)

**Review by joelvdavies — COMMENTED** [review]
One minor thing and same query about tests/linting.

Also note for @joshuadkitenge, it appears operations gateway-api/this actually currently uses the icat provider instead of the jwtAuthProvider (https://github.com/ral-facilities/scigateway/blob/8c501b09aeec1fdf6c04474c080b1b30577c86be/src/authentication/icatAuthProvider.tsx vs https://github.com/ral-facilities/scigateway/blob/8c501b09aeec1fdf6c04474c080b1b30577c86be/src/authentication/jwtAuthProvider.tsx). The main difference is the inclusion of the maintenance states by the looks of it. Also the jwt provider instead uses `/api/jwt/authenticate`, `/api/jwt/checkToken`, `/api/jwt/refresh` instead which will also not currently work unless we change this to use that instead.

File: ldap_jwt_auth/routers/refresh.py (hunk 2 above)
Comment (VKTB) [reply ldap_jwt_auth/routers/refresh.py]: Addressed in https://github.com/ral-facilities/ldap-jwt-auth/pull/11/commits/886f9c1f5115a369688a0d31c3c151566bc6f93a

**Review by joelvdavies — COMMENTED** [review]
Code LGTM - On the refresh issue, it appears SciGateway is missing `{ withCredentials: true }` in the refresh axios request to send the cookie - Adding it made it work for me otherwise it couldn't find the cookie like you mentioned. Although I don't know why this hasn't already been addressed unless it hasn't been tried yet or there is some other way around it.

Although even with that change - it fails as the token is returned in the body but needs
```
{
    "token": token
}
```
instead of
```
{
    token
}
```
I don't know why it is different to the login though - it doesn't look like operations gateway does this either, so I suspect SciGateway needs updating.

**Comment by VKTB** [comment]
@joelvdavies I am not sure about the cookie, sorry. Regarding the response for the `/refresh` endpoint, SciGateway Auth just returns a string like the `/login` endpoint does.

**Comment by joelvdavies** [comment]
@VKTB SciGateway auth also has a maintenance endpoint so it looks like it uses SciGateway's `icatAuthProvider`, if I look at https://github.com/ral-facilities/scigateway/pull/1275/files#diff-262fc2f51a09abb09571717b6aa50067596d5fd00a8953bf8b1624ad2ce4f281R15 - The login was changed to use data.token, but the refresh was left as `res.data.token`. I will raise it with Louise as I have a catch-up today anyway, it definitely looks like it just needs changing too.

**Review by joelvdavies — APPROVED** [review]
It seems like the issue with refresh is likely CORS related, so should be solved when going via proxy/potentially by using https so have agreed to test this again once its setup.

**Pull request merged**